core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala

```diff
@@ -398,40 +414,96 @@
     }
 
 
-  final def updateWorkspaceRecord: MigrateAction[Unit] =
+  final def restoreIamPoliciesAndUpdateWorkspaceRecord: MigrateAction[Unit] =
     withMigration(_.tmpBucketDeleted.isDefined) { (migration, workspace) =>
       for {
-        _ <- MigrateAction.raiseWhen(migration.newGoogleProjectId.isEmpty || migration.newGoogleProjectNumber.isEmpty) {
-          noGoogleProjectError(migration, workspace)
+        (workspaceService, sam, userInfo) <- MigrateAction.asks { env =>
+          (env.workspaceService, env.sam, env.userInfo)
         }
-        _ <- inTransaction(dataAccess =>
-          dataAccess
-            .workspaceQuery
+
+        googleProjectId = migration.newGoogleProjectId.getOrElse(
+          throw noGoogleProjectError(migration, workspace)
+        )
+
+        _ <- MigrateAction.fromFuture {
```

Comment:
Our team has sort of settled on a best practice of trying to avoid nested `for` comps and/or long logical blocks like this one for the sake of readability and maintainability.  It's not a law, but something we try to keep an eye out for.  Suggest extracting the right side of this "gets"  (I think that's what you referred to these statements as?) into its own method called something like, `grantPermissionsOnGoogleProjectForBillingProjectOwners` or something along those lines.

Reply:
I think these nested `for`s are going to change in the next PR and so pulling them out into separate functions is going to lead to more pain than what it's worth. We can do a tidy-up PR when all this is finalised.